### PR TITLE
Resolve and init modules mentioned on command line

### PR DESCRIPTION
**Pull request by mppf**
For #18714

Related to PR #13369

This PR
 * resolves and arranges to run module init/deinit for any top-level module mentioned on the command line
 * changes to setting FLAG_MODULE_FROM_COMMAND_LINE_FILE only for top-level modules mentioned on the command line (not for nested modules). To keep main-module inference working I needed to add a helper function isModuleOrSubmoduleFromCommandLine in ModuleSymbol.cpp.
 * removes currentFileNamedOnCommandLine and instead takes the approach of setting on top-level modules parsed from command line files after parsing.
 * updates a comment about how to add a flag since the flags are now in the dyno library

Reviewed by @lydia-duncan - thanks!

- [x] full local testing

### CODE REVIEW

**Comment by bradcray** [comment]
I'm seeing 184 errors on this branch, most of which exhibit as internal errors of one form or another with the LLVM back-end.  E.g., for `test/classes/initializers/secondary/inDiffModule.chpl`, the LLVM back-end gets an access error (`internal error: COD-CG--BOL-2765 chpl version 1.26.0 pre-release (88ac3197ab)
`) at the following stack trace:

```
  * frame #0: 0x000000010016b037 chpl`Symbol::isRef(this=0x0000000111b840f0) at symbol.cpp:236:21
    frame #1: 0x0000000100073e90 chpl`pruneVisit(sym=0x0000000111b840f0, fns=0x00007ffeefbff2f0, types=0x00007ffeefbff2c0) at astutil.cpp:858:14
    frame #2: 0x00000001000741c7 chpl`pruneVisitFn(fn=0x0000000113e23ed0, fns=0x00007ffeefbff2f0, types=0x00007ffeefbff2c0) at astutil.cpp:891:7
    frame #3: 0x0000000100073e69 chpl`pruneVisit(sym=0x0000000113e23ed0, fns=0x00007ffeefbff2f0, types=0x00007ffeefbff2c0) at astutil.cpp:855:5
    frame #4: 0x000000010006cc5e chpl`visitVisibleFunctions(fns=0x00007ffeefbff2f0, types=0x00007ffeefbff2c0) at astutil.cpp:906:5
    frame #5: 0x000000010006ca3f chpl`prune() at astutil.cpp:1105:3
    frame #6: 0x000000010006d469 chpl`prune2() at astutil.cpp:1122:17
    frame #7: 0x000000010036e840 chpl`runPass(tracker=0x00007ffeefbff3e0, passIndex=33, isChpldoc=false) at runpasses.cpp:208:3
    frame #8: 0x000000010036e6bd chpl`runPasses(tracker=0x00007ffeefbff3e0, isChpldoc=false) at runpasses.cpp:170:5
    frame #9: 0x0000000100360987 chpl`main(argc=5, argv=0x00007ffeefbff458) at driver.cpp:1756:3
 ```

The C back-end seems to complain about `chpl__init_B()` not having a prototype in the generated code (and is one of the cases I feel like I've been seeing lately where the compiler doesn't generate a non-zero exit code even though the C compilation broke, so I'm looking into that next).  Given the nature of the test, it makes me wonder whether something more needs to be done to handle the case of two modules defined within a single file on the command-line.

These failures were fishy enough, and the others similar enough on the surface, that I didn't look at other failure cases much.

**Comment by mppf** [comment]
Thanks, I'll have a look and see if I can fix inDiffModule.chpl. I can see it has --verify errors currently.

**Comment by mppf** [comment]
I have a fix for inDiffModule and I'm going to go ahead and launch a test run.

**Comment by mppf** [comment]
Here is the test result with --verify
```
[Error matching compiler output for constrained-generics/basic/set1/error-cg-arg]
[Error matching program output for extern/bradc/require/useModAsString]
[Error matching compiler output for extern/require/requireInSubMod]
[Error matching compiler output for extern/require/requireInTopMod]
[Error matching program output for modules/diten/main-module-flag/multiple_modules_no_main2]
[Error matching program output for modules/diten/main-module-flag/multiple_modules_no_main3]
[Error matching program output for modules/errors/code-outside-of-module-error]
[Error matching program output for modules/noakes/mod-init-fails]
[Error matching compiler output for modules/sungeun/resolution/module-resolution-issue (compopts: 1)]
[Error matching program output for modules/sungeun/resolution/module-resolution-issue (compopts: 2)]
[Error matching compiler output for modules/sungeun/no-use-class]
[Error matching compiler output for modules/sungeun/no-use-enum]
[Error matching compiler output for modules/sungeun/no-use-record]
[Error matching compiler output for modules/sungeun/no-use-type-alias]
[Error matching program output for optimizations/deadCodeElimination/elliot/deadModuleElim]
[Error matching program output for performance/vectorization/vectorPragmas/loopsInForallNoVector]
[Error matching program output for performance/vectorization/vectorPragmas/zipIterNoVector]
[Error matching compiler output for release/examples/primers/errorHandling]
[Error matching program output for release/examples/primers/modules]
[Summary: #Successes = 13261 | #Failures = 19 | #Futures = 0]
```

**Comment by bradcray** [comment]
error-cg-arg raises an interesting question about whether nested modules in files specified on the command-line should also be initialized even if they aren't used.  I don't have an immediate intuition about that question, but if I had to pick a side quickly, I'd probably say "no."

**Comment by bradcray** [comment]
With a quick run-through of the rest of the tests, it looked like 1/2 - 1/3 were other cases of nested modules.  The noakes, sungeun, elliot, and primers looked like they'd require another look.

**Comment by bradcray** [comment]
Taking a look, it seems that most of these failures are "expected" due to the change, with the exception of a pair of internal errors and two tests that I didn't yet sort out what was happening.  Bottom line, this looks close to me:

> [Error matching compiler output for extern/require/requireInTopMod]

This was checking that a non-used module didn't affect requirements, but now it will, as it arguably should.

> [Error matching compiler output for modules/sungeun/no-use-class]
> [Error matching compiler output for modules/sungeun/no-use-enum]
> [Error matching compiler output for modules/sungeun/no-use-record]
> [Error matching compiler output for modules/sungeun/no-use-type-alias]
> [Error matching program output for modules/sungeun/resolution/module-resolution-issue (compopts: 2)]

These all seem to be checking that an unused module doesn't get compiled, but not it will be, so new errors are generated.

> [Error matching program output for optimizations/deadCodeElimination/elliot/deadModuleElim]

New dead modules are eliminated because they were considered more than usual?

> [Error matching compiler output for modules/noakes/mod-init-fails]
> [Error matching compiler output for modules/sungeun/resolution/module-resolution-issue (compopts: 1)]

Internal error—needs a look

> [Error matching program output for performance/vectorization/vectorPragmas/loopsInForallNoVector]
> [Error matching program output for performance/vectorization/vectorPragmas/zipIterNoVector]

Wasn't obvious to me what was happening here.

File: compiler/AST/ModuleSymbol.cpp [hunk 1]
```diff
@@ -134,6 +134,17 @@ ModuleSymbol* ModuleSymbol::findMainModuleByName() {
   return retval;
 }
 
+static bool isModuleOrSubmoduleFromCommandLine(ModuleSymbol* mod) {
```
Comment (lydia-duncan) [inline compiler/AST/ModuleSymbol.cpp:137]: I think you could shorten this to `isModOrSubmodFromCommandLine`

File: compiler/include/flags.h [hunk 2]
```diff
@@ -24,7 +24,8 @@
 
 //
 // HOW TO ADD A FLAG:
-// - add an entry (sorted alphabetically) to the list in flags_list.h
+// - add an entry (sorted alphabetically) to the list in
+//   compiler/dyno/include/chpl/uast/PragmaList.h
```
Comment (lydia-duncan) [inline compiler/include/flags.h:28]: Was this intentionally included in this PR?  I think it should be mentioned in the PR message.

File: compiler/optimizations/deadCodeElimination.cpp [hunk 3]
```diff
@@ -355,6 +355,11 @@ static bool isDeadModule(ModuleSymbol* mod) {
   if (mod == ModuleSymbol::mainModule())
     return false;
 
+  // Modules named on the command line should not be considered dead
+  // either since the init function should be called.
+  if (mod->hasFlag(FLAG_MODULE_FROM_COMMAND_LINE_FILE))
```
Comment (lydia-duncan) [inline compiler/optimizations/deadCodeElimination.cpp:360]: Will deadCodeElimination get rid of the submodules?

File: test/extern/require/requireInTopMod.chpl [hunk 4]
```diff
@@ -1,11 +1,8 @@
 module M {
-  //
-  // The following file intentionally contains syntax errors to make
-  // sure we don't add requirements for modules that are not 'use'd.
-  //
-  require "bar.h";
```
Comment (lydia-duncan) [inline test/extern/require/requireInTopMod.chpl:6]: Is `bar.h` used in other programs in that directory?  Should it be removed?

**Review by lydia-duncan — APPROVED** [review]
Just a couple of suggestions but nothing major

File: compiler/optimizations/deadCodeElimination.cpp (hunk 3 above)
Comment (mppf) [reply compiler/optimizations/deadCodeElimination.cpp:360]: Yes, and we have tests that demonstrate that

File: test/extern/require/requireInTopMod.chpl (hunk 4 above)
Comment (mppf) [reply test/extern/require/requireInTopMod.chpl:6]: requireInSubMod.chpl uses it.
